src/kernel/lockstats.c

```diff
@@ -0,0 +1,312 @@
+#include <kernel.h>
+#include <net.h>
+#include <http.h>
+#include <symtab.h>
+
+#define LOCKSTATS_PORT               9090
+#define LOCKSTATS_URI                "lockstat"
+#define LOCKSTATS_HTTP_CHUNK_MAXSIZE (64*KB)
+#define LOCKSTATS_PREALLOC           2048
+
+static heap lockstats_heap;
+static http_listener lockstats_hl;
+
+boolean record_lock_stats = false;
+
+static u64 hash_lock_block(lock_block b)
+{
+    u64 hash = 0xcbf29ce484222325;
+    u64 fnv_prime = 1099511628211;
+    for (int i = 0; i < MAX_TRACE_DEPTH; i++) {
+        if (b->lock_trace[i] == 0)
+            break;
+        hash ^= b->lock_trace[i];
+        hash *= fnv_prime;
+    }
+    return hash;
+}
+
+static inline void save_frame_trace(u64 *trace)
+{
+    int i;
+    u64 *fp = get_current_fp();
+
+    for (i = 0; i < MAX_TRACE_DEPTH; i++) {
+        u64 *rap = get_frame_ra_ptr(fp, &fp);
+        if (rap == 0)
+            break;
+        trace[i] = *rap;
+        if (*rap == 0)
+            break;
+    }
+    if (i < MAX_TRACE_DEPTH)
+        trace[i] = 0;
+}
+
+lock_stats get_lockstats_block(lockstats_lock l, boolean islocking)
+{
+    struct lock_block lookup;
+    cpuinfo ci = current_cpu();
+
+    if (ci->lock_stats_disable)
+        return 0;
+    ci->lock_stats_disable = true;
+
+    u64 trace_hash = 0;
+    lock_stats stats = 0;
+    table t = ci->lock_stats_table;
+    if (islocking) {
+        lookup.lock_address = u64_from_pointer(l);
+        save_frame_trace(lookup.lock_trace);
+        trace_hash = hash_lock_block(&lookup);
+        stats = table_find(t, (void *)trace_hash);
+        if (stats)
+            goto out;
+        stats = allocate_zero(ci->lock_stats_heap, sizeof(struct lock_stats));
+        assert(stats != INVALID_ADDRESS);
+        runtime_memcpy(&stats->lock, &lookup, sizeof(struct lock_block));
+        stats->lock.type = l->type;
+        stats->lock.hash = trace_hash;
+        stats->spins_min = -1u;
+        stats->hold_time_min = -1u;
+        table_set(t, (void *)trace_hash, stats);
+    } else if (l->trace_hash) {
+        stats = table_find(t, (void *)l->trace_hash);
+    }
+out:
+    ci->lock_stats_disable = false;
+    return stats;
+}
+
+boolean lockstats_print_u64_with_sym(buffer b, u64 n)
+{
+    char * name;
+    u64 offset, len;
+
+    name = find_elf_sym(n, &offset, &len);
+    bprintf(b, "%p [%s+0x%x]", n, name ? name : "", name ? offset : 0);
+    return name != 0;
+}
+
+static boolean stat_sort_reverse(void *a, void *b)
+{
+    lock_stats sa = a;
+    lock_stats sb = b;
+    if (sa->cont > sb->cont)
+        return false;
+    return true;
+}
+
+static pqueue log_collate_and_sort(void)
+{
+    cpuinfo ci;
+    record_lock_stats = 0;
```

Comment:
Should probably check the current value of `record_lock_stats` and restore it at the end of this function.